core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -867,41 +893,19 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
         )
       }
     }
-
-  def memoize: ZManaged[R, E, ZManaged[R, E, A]] =
-    ZManaged {
-      RefM.make[Option[(Reservation[R, E, A], Exit[E, A])]](None).map { ref =>
-        val acquire1: ZIO[R, E, A] =
-          ref.modify {
-            case v @ Some((_, e)) => ZIO.succeed(e -> v)
-            case None =>
-              ZIO.uninterruptibleMask { restore =>
-                self.reserve.flatMap { res =>
-                  restore(res.acquire).run.map(e => e -> Some(res -> e))
-                }
-              }
-          }.flatMap(ZIO.done)
-
-        val acquire2: ZIO[R, E, ZManaged[R, E, A]] =
-          ZIO.succeed(acquire1.toManaged_)
-
-        val release2 = (_: Exit[_, _]) =>
-          ref.updateSome {
-            case Some((res, e)) => res.release(e).as(None)
-          }
-
-        Reservation(acquire2, release2)
-      }
-    }
 }
 
 object ZManaged {
 
-  /**
-   * Creates new [[ZManaged]] from wrapped [[Reservation]].
-   */
-  final def apply[R, E, A](reservation: ZIO[R, E, Reservation[R, E, A]]): ZManaged[R, E, A] =
-    new ZManaged(reservation)
+  final class AccessPartiallyApplied[R](private val dummy: Boolean = true) extends AnyVal {
+    def apply[A](f: R => A): ZManaged[R, Nothing, A] =
+      ZManaged.environment.map(f)
+  }
+
+  final class AccessMPartiallyApplied[R](private val dummy: Boolean = true) extends AnyVal {
+    def apply[E, A](f: R => ZManaged[R, E, A]): ZManaged[R, E, A] =
```

Comment:
Should `f` here be `R => ZManaged[R, E, A]` or `R => ZIO[R, E, A]`? I ran into the same thing with regard to `accessM` on `ZStream`. It seems like it could just be the first one, and that is more consistent with other `M` methods, but I'm not sure.

Reply:
My feeling is that we are more flexible with the current signature.
I think the proposed signature would be equivalent to ZIO.accessM(???).toManaged_ which I feel is something you need less often.

Reply:
Added a comment related to this in #2337.

Reply:
I'll rename this to accessManaged and add accessM as per John's comment

Reply:
Sounds good. I am doing the same thing on `ZStream`.